mysql/datadog_checks/mysql/activity.py

```diff
@@ -0,0 +1,292 @@
+import datetime
+import decimal
+import time
+from contextlib import closing
+from enum import Enum
+from typing import Dict, List
+
+import pymysql
+
+from datadog_checks.base import is_affirmative, to_native_string
+from datadog_checks.base.utils.db.sql import compute_sql_signature
+from datadog_checks.base.utils.db.utils import DBMAsyncJob, obfuscate_sql_with_metadata
+from datadog_checks.base.utils.serialization import json
+from datadog_checks.base.utils.tracking import tracked_method
+
+try:
+    import datadog_agent
+except ImportError:
+    from ..stubs import datadog_agent
+
+CONNECTIONS_QUERY = """\
+SELECT
+    processlist_user,
+    processlist_host,
+    processlist_db,
+    processlist_state,
+    COUNT(processlist_user) AS connections
+FROM
+    performance_schema.threads
+WHERE
+    processlist_user IS NOT NULL AND
+    processlist_state IS NOT NULL
+    GROUP BY processlist_user, processlist_host, processlist_db, processlist_state
+"""
+
+ACTIVITY_QUERY = """\
+SELECT
+    thread_a.thread_id,
+    thread_a.processlist_id,
+    thread_a.processlist_user,
+    thread_a.processlist_host,
+    thread_a.processlist_db,
+    thread_a.processlist_command,
+    thread_a.processlist_state,
+    thread_a.processlist_info AS sql_text,
+    statement.timer_start AS event_timer_start,
+    statement.timer_end AS event_timer_end,
+    statement.lock_time,
+    statement.current_schema,
+    COALESCE(
+        IF(thread_a.processlist_state = 'User sleep', 'User sleep',
+        IF(waits_a.event_id = waits_a.end_event_id, 'CPU', waits_a.event_name)), 'CPU') AS wait_event,
+    waits_a.event_id,
+    waits_a.end_event_id,
+    waits_a.event_name,
+    waits_a.timer_start AS wait_timer_start,
+    waits_a.timer_end AS wait_timer_end,
+    waits_a.object_schema,
+    waits_a.object_name,
+    waits_a.index_name,
+    waits_a.object_type,
+    waits_a.source,
+    socket.ip,
+    socket.port
+FROM
+    performance_schema.threads AS thread_a
+    LEFT JOIN performance_schema.events_waits_current AS waits_a ON waits_a.thread_id = thread_a.thread_id AND
+    waits_a.event_id IN(
+        SELECT
+            MAX(waits_b.EVENT_ID)
```

Comment:
What is this doing exactly?

Reply:
the view events_waits_current can have multiple rows per thread (despite the documentation that states otherwise), thus we use EVENT_ID to identify the row we want to use. We want the row with the highest EVENT_ID value as the most recent and reflects the current wait

Reply:
Very interesting! That might be worth documenting for future generations of maintainers...